core/commonKotlin/src/Instant.kt

```diff
@@ -168,6 +169,266 @@ public actual class Instant internal constructor(public actual val epochSeconds:
+            // org.threeten.bp.LocalDate#toEpochDay
+            run {
+                var zeroDay = epochDays + DAYS_0000_TO_1970
+                // find the march-based year
+                zeroDay -= 60 // adjust to 0000-03-01 so leap day is at end of four year cycle
+
+                var adjust = 0L
+                if (zeroDay < 0) { // adjust negative years to positive for calculation
+                    val adjustCycles = (zeroDay + 1) / DAYS_PER_CYCLE - 1
+                    adjust = adjustCycles * 400
+                    zeroDay += -adjustCycles * DAYS_PER_CYCLE
+                }
+                var yearEst = ((400 * zeroDay.toLong() + 591) / DAYS_PER_CYCLE)
+                var doyEst = zeroDay - (365 * yearEst + yearEst / 4 - yearEst / 100 + yearEst / 400)
+                if (doyEst < 0) { // fix estimate
+                    yearEst--
+                    doyEst = zeroDay - (365 * yearEst + yearEst / 4 - yearEst / 100 + yearEst / 400)
+                }
+                yearEst += adjust // reset any negative year
+
+                val marchDoy0 = doyEst.toInt()
+
+                // convert march-based values back to january-based
+                val marchMonth0 = (marchDoy0 * 5 + 2) / 153
+                month = (marchMonth0 + 2) % 12 + 1
+                day = marchDoy0 - (marchMonth0 * 306 + 5) / 10 + 1
+                year = (yearEst + marchMonth0 / 10).toInt()
+            }
+            val hours = (secsOfDay / SECONDS_PER_HOUR)
+            val secondWithoutHours = secsOfDay - hours * SECONDS_PER_HOUR
+            val minutes = (secondWithoutHours / SECONDS_PER_MINUTE)
+            val second = secondWithoutHours - minutes * SECONDS_PER_MINUTE
+            return UnboundedLocalDateTime(year, month, day, hours, minutes, second, instant.nanosecondsOfSecond)
+        }
+    }
+}
+
+internal fun parseIso(isoString: String): Instant {
+    fun parseFailure(error: String): Nothing {
+        throw IllegalArgumentException("$error when parsing an Instant from $isoString")
+    }
+    inline fun expect(what: String, where: Int, predicate: (Char) -> Boolean) {
+        val c = isoString[where]
+        if (!predicate(c)) {
+            parseFailure("Expected $what, but got $c at position $where")
+        }
+    }
+    val s = isoString
+    var i = 0
+    require(s.isNotEmpty()) { "An empty string is not a valid Instant" }
+    val yearSign = when (val c = s[i]) {
+        '+', '-' -> { ++i; c }
+        else -> ' '
+    }
+    val yearStart = i
+    var absYear = 0
+    while (i < s.length && s[i] in '0'..'9') {
+        absYear = absYear * 10 + (s[i] - '0')
+        ++i
+    }
+    val year = when {
+        i > yearStart + 10 -> {
+            parseFailure("Expected at most 10 digits for the year number, got ${i - yearStart}")
+        }
+        i == yearStart + 10 && s[yearStart] >= '2' -> {
+            parseFailure("Expected at most 9 digits for the year number or year 1000000000, got ${i - yearStart}")
```

Comment:
Minor: can be unclear from the error message that `i - yearStart` is the actual number of digits